File: python/src/azoth/eos/reference/_hydrate.py
"""The van der Waals-Platteeuw hydrate fugacity: the Python reference.

The mirror of ``crates/azoth-eos/src/hydrate.rs``, which states the physics and why the
route is the fitted one - NeqSim's ``ComponentHydratePVTsim``, whose guest constants are
``C = A/T exp(B/T)`` from the ``HydrateA*``/``B*`` columns, rather than the base class's
Kihara integral over the three ``*HYDRATE`` Lennard-Jones columns. It also states the two
places a port goes wrong silently: the ``C f`` product is taken in **bar**, and the
coefficient's ``f_w^ref`` is a *computed* fugacity rather than the pressure.
"""

from __future__ import annotations

import math

from azoth.core.errors import InvalidInputError, OutOfRangeError

#: NeqSim's ``ThermodynamicConstantsInterface.R``, which ``calcDeltaChemPot`` divides by.
R = 8.3144621

#: The reference temperature of the chemical-potential constant, K.
T_REFERENCE = 273.15

#: How many cavities of each type a water molecule belongs to: structure I then II, small
#: then large. NeqSim's ``ComponentHydrate`` constructor, and hardcoded rather than read.
CAVITIES_PER_WATER: tuple[tuple[float, float], tuple[float, float]] = (
    (1.0 / 23.0, 3.0 / 23.0),
    (2.0 / 17.0, 1.0 / 17.0),
)

#: The chemical-potential constants ``(dGf, dHf, Cp, dV)`` of ``calcDeltaChemPot``,
#: structure I then II, in J/mol, J/mol, J/(mol K) and m3/mol.
_CHEMICAL_POTENTIAL: tuple[tuple[float, float, float, float], ...] = (
    (1264.0, -4858.0, -39.16, 4.6e-6),
    (883.0, -5201.0, -39.16, 5.0e-6),
)

#: NeqSim works its fugacities in bara and this states its own in pascals.
_PA_PER_BAR = 1.0e5

#: The reference pressure the empty-lattice vapour pressure is stated against, bar. NeqSim's
#: ``ThermodynamicConstantsInterface.referencePressure``.
_REFERENCE_PRESSURE_BAR = 1.01325

#: The empty structure's vapour pressure constants, structure I then II: Sloan (1990), as
#: NeqSim's ``ComponentHydrate.emptyHydrateVapourPressureConstant``.
_EMPTY_VAPOUR_PRESSURE = ((17.44, -6003.9), (17.332, -6017.6))

#: Avlonitis (1994)'s molar volume of the empty hydrate, ``(v0, k1, k2, k3)`` in cm3/mol and
#: per K, structure I then II. NeqSim's ``ComponentHydrate.getMolarVolumeHydrate``.
_HYDRATE_VOLUME = (
    (22.35, 3.1075e-4, 5.9537e-7, 1.3707e-10),
    (22.57, 1.9335e-4, 2.1768e-7, -1.4786e-10),
)

#: Which fitted hydrate model a calculation runs. NeqSim picks between them by the *system's
#: model name* rather than by anything about the fluid, so the name is a model-level choice.
PVTSIM = "pvtsim"
GUO_FINCH = "guo_finch"

#: The cavities of each type in one unit cell: structure I then II, small then large.
#: NeqSim's ``ComponentHydrate`` constructor, and the counts its ``46/54`` and ``136/160``
#: bounds come from - which are the *fully occupied* limits, not the composition at a state.
CAVITIES_PER_CELL: tuple[tuple[float, float], tuple[float, float]] = ((2.0, 6.0), (16.0, 8.0))

#: The water molecules in one unit cell.
WATER_PER_CELL: tuple[float, float] = (46.0, 136.0)


class HydrateGuest:
    """A substance's hydrate record: what the occupancy loops read, and no more."""

    __slots__ = ("former", "guo_finch_a", "guo_finch_b", "langmuir_a", "langmuir_b", "name")

    def __init__(
        self,
        name: str,
        langmuir_a: tuple[float, float, float, float],
        langmuir_b: tuple[float, float, float, float],
        guo_finch_a: tuple[float, float, float, float],
        guo_finch_b: tuple[float, float, float, float],
        former: bool,
    ) -> None:
        self.name = name
        self.langmuir_a = langmuir_a
        self.langmuir_b = langmuir_b
        self.guo_finch_a = guo_finch_a
        self.guo_finch_b = guo_finch_b
        self.former = former


class Hydration:
    """What a mixture needs to have a hydrate calculated for it."""

    __slots__ = ("guests", "water_index")

    def __init__(self, guests: list[HydrateGuest], water_index: int | None) -> None:
        self.guests = guests
        self.water_index = water_index


def langmuir(guest: HydrateGuest, model: str, structure: int, cavity: int, t: float) -> float:
    """``C = A/T exp(B/T)`` for one guest at one structure's cavity, on one model's pair."""
    index = structure * 2 + cavity
    a, b = (
        (guest.langmuir_a, guest.langmuir_b)
        if model == PVTSIM
        else (guest.guo_finch_a, guest.guo_finch_b)
    )
    return a[index] / t * math.exp(b[index] / t)


def empty_hydrate_vapour_pressure(structure: int, t: float) -> float:
    """The empty hydrate's vapour pressure on one structure, **in bar**.

    The Guo-Finch route's reference term, and the reason that route breaks the structure
    comparison: it is a function of the structure, so it does not cancel the way the PVTsim
    route's single reference does.
    """
    c0, c1 = _EMPTY_VAPOUR_PRESSURE[structure]
    return math.exp(c0 + c1 / t) * _REFERENCE_PRESSURE_BAR


def molar_volume_hydrate(structure: int, t: float) -> float:
    """The molar volume of the empty hydrate on one structure, m3/mol. Avlonitis (1994)."""
    v0, k1, k2, k3 = _HYDRATE_VOLUME[structure]
    dt = t - T_REFERENCE
    return v0 * (1.0 + k1 * dt + k2 * dt * dt + k3 * dt * dt * dt) / 1.0e6


def _reference_term(
    model: str, structure: int, t: float, p: float, reference_water_fugacity: float
) -> float:
    """What multiplies ``exp(exponent)`` to make the coefficient."""
    if model == PVTSIM:
        return reference_water_fugacity / p
    p_empty = empty_hydrate_vapour_pressure(structure, t) * _PA_PER_BAR
    return p_empty * math.exp(molar_volume_hydrate(structure, t) / (R * t) * (p - p_empty)) / p


def occupancy(
    guests: list[HydrateGuest],
    ref_fugacities: list[float],
    model: str,
    structure: int,
    cavity: int,
    t: float,
) -> list[float]:
    """One cavity type's occupancy by each guest, ``C f_i / (1 + sum_j C_j f_j)``.

    **The product ``C f`` is taken in bar**, whatever unit the caller states its fugacities
    in: the Langmuir constant is fitted against NeqSim's bar-valued ones and carries the
    reciprocal. Feeding it pascals saturates every cage.
    """
    denominator = 1.0
    for guest, fugacity in zip(guests, ref_fugacities, strict=True):
        if guest.former:
            denominator += langmuir(guest, model, structure, cavity, t) * fugacity / _PA_PER_BAR
    return [
        langmuir(guest, model, structure, cavity, t) * fugacity / _PA_PER_BAR / denominator
        if guest.former
        else 0.0
        for guest, fugacity in zip(guests, ref_fugacities, strict=True)
    ]


def _chemical_potential(structure: int, t: float, p: float) -> float:
    """``calcDeltaChemPot``: the change in chemical potential on forming the hydrate."""
    dgf, dhf, cp, dvolume = _CHEMICAL_POTENTIAL[structure]
    return (
        dgf / R / T_REFERENCE
        - (
            -dhf * (1.0 / R / t - 1.0 / R / T_REFERENCE)
            + cp / R * math.log(t / T_REFERENCE)
            + cp * T_REFERENCE / R * (1.0 / t - 1.0 / T_REFERENCE)
        )
        + dvolume / R / t * p
    )


def exponent(
    guests: list[HydrateGuest],
    ref_fugacities: list[float],
    model: str,
    t: float,
    p: float,
    structure: int,
) -> float:
    """The exponent of a structure's water fugacity coefficient: the part that is its own.

    ``f_w^hydrate/P = (f_w^ref/P) exp(sum_cav n_cav ln(1 - sum_j theta_j) + dMu)``.

    **The fluid's own water fugacity cancels out of this**, which is why it is not an
    argument. NeqSim writes the coefficient as
    ``(f_w^fluid/P) exp(sum + dMu + ln(f_w^ref/f_w^fluid))``, where the fluid's water fugacity
    is multiplied in and divided out again inside the logarithm - it is not what the
    hydrate's water fugacity depends on, and ``f_w^ref`` is. Written out that way the
    coefficient is ``0 * inf`` on a fluid with no water, which is a state a hydrate
    fraction's *bound* is, so the cancellation is taken here.

    Raises:
        OutOfRangeError: if a cavity type is fully occupied, where the sum's logarithm has no
            value. A refusal rather than a clamp: a saturation of one means the occupancy
            model has left the region it was fitted to.
    """
    val = 0.0
    for cavity, per_water in enumerate(CAVITIES_PER_WATER[structure]):
        occupied = sum(occupancy(guests, ref_fugacities, model, structure, cavity, t))
        if occupied >= 1.0:
            raise OutOfRangeError(
                "occupancy",
                occupied,
                f"cavity {cavity} of structure {structure + 1} is fully occupied at {t} K, "
                f"where the cavity sum's logarithm has no value",
            )
        val += per_water * math.log(1.0 - occupied)
    # The Guo-Finch route's reference is the empty lattice's own vapour pressure, carried by
    # `_reference_term`, so it adds no chemical-potential constant.
    return val + (_chemical_potential(structure, t, p) if model == PVTSIM else 0.0)


def water_fugacity_coefficient(
    guests: list[HydrateGuest],
    ref_fugacities: list[float],
    model: str,
    t: float,
    p: float,
    structure: int,
    reference_water_fugacity: float,
) -> float:
    """The hydrate's water fugacity coefficient on one structure, ``f_w^hydrate/P``.

    ``f_w^ref`` is the reference water phase's fugacity and is **not** the pressure: NeqSim
    builds it from a one-component phase of the host's own class, so its water component is
    the host's and its fugacity is a real number. The caller states it, because only the
    caller knows which equation the fluid is.

    Raises:
        OutOfRangeError: from the structure's cavity sum.
    """
    return math.exp(exponent(guests, ref_fugacities, model, t, p, structure)) * _reference_term(
        model, structure, t, p, reference_water_fugacity
    )


def stable_structure(
    guests: list[HydrateGuest],
    ref_fugacities: list[float],
    model: str,
    t: float,
    p: float,
    reference_water_fugacity: float,
) -> tuple[int, float]:
    """The stable structure and its coefficient: the **lower** of the two.

    **The coefficients are compared, not the exponents.** The PVTsim route's reference is one
    number for both structures, so comparing exponents happens to agree there; the Guo-Finch
    route's is the empty lattice's vapour pressure on that structure, which does not cancel.
    """
    first = water_fugacity_coefficient(
        guests, ref_fugacities, model, t, p, 0, reference_water_fugacity
    )
    second = water_fugacity_coefficient(
        guests, ref_fugacities, model, t, p, 1, reference_water_fugacity
    )
    return (0, first) if first <= second else (1, second)


def composition(
    guests: list[HydrateGuest],
    ref_fugacities: list[float],
    model: str,
    structure: int,
    t: float,
    water_index: int,
) -> list[float]:
    """The hydrate's mole fractions at a state, from **both** cavity types of a structure.

    A cell of structure I holds 46 waters with two small and six large cavities, and one of
    structure II 136 with sixteen and eight. A cavity of type ``cav`` holds guest ``i`` with
    probability ``theta_icav``, so the cell's guest count is the cavities' own weighted sum
    of the occupancies, and each guest's is the one weighted by its own.

    **The second cavity is the whole point.** A guest occupies one cage type at a time and
    its occupancy in each is a different function of the state, so a composition built from one
    cavity type leaves the other's guests at zero. For structure I that is the large
    ``5^12 6^2`` cage, where most of the ethane and propane sit: the phase's fractions would not
    sum to one, and its bound would not move with the temperature.
    """
    counts = [0.0] * len(guests)
    total_guests = 0.0
    for cavity in range(2):
        per_cell = CAVITIES_PER_CELL[structure][cavity]
        for index, occupied in enumerate(
            occupancy(guests, ref_fugacities, model, structure, cavity, t)
        ):
            counts[index] += per_cell * occupied
            total_guests += per_cell * occupied
    water = WATER_PER_CELL[structure]
    total = water + total_guests
    fractions = [count / total for count in counts]
    fractions[water_index] = water / total
    return fractions


def hydration_for(names: list[str]) -> Hydration:
    """The guest records for a set of names, refused if the fluid cannot form a hydrate.

    Raises:
        InvalidInputError: if a name is not in the databank, if there is no water, or if
            nothing in the mixture is a hydrate former.
    """
    from azoth.eos import components as databank

    guests = []
    for name in names:
        entry = databank.entry(name)
        guests.append(
            HydrateGuest(
                name=entry.name,
                langmuir_a=entry.hydrate_langmuir_a,
                langmuir_b=entry.hydrate_langmuir_b,
                guo_finch_a=entry.hydrate_guo_finch_a,
                guo_finch_b=entry.hydrate_guo_finch_b,
                former=entry.hydrate_former,
            )
        )

    water_index = next((i for i, g in enumerate(guests) if g.name == "water"), None)
    if water_index is None:
        raise InvalidInputError(
            "components",
            "no water: a hydrate is water's, so a fluid without it has no formation temperature",
        )
    if not any(guest.former for guest in guests):
        raise InvalidInputError(
            "components",
            "nothing in this mixture is a hydrate former, so no cage would have a guest and "
            "the hydrate's fugacity would be its empty one",
        )
    return Hydration(guests=guests, water_index=water_index)

File: crates/azoth-eos/tests/hydrate.rs
//! The hydrate fugacity, checked against the capture's own layers.
//!
//! `validation/neqsim/captures/hydrate_probe.tsv` prints the gas phase's per-component
//! fugacities at the converged state, which are exactly what NeqSim's `setFug` copies into
//! the hydrate's `reffug`. So the arithmetic can be driven from those four numbers alone and
//! compared with the occupancies and the water fugacity the same capture reports - no fluid
//! flash, and therefore no way for a fluid-side difference to hide a hydrate-side one.
//!
//! **The structure is II at every one of these states, and the capture said I for a while.**
//! The stable structure lives on the *water* component: `ComponentHydratePVTsim.fugcoef` runs
//! its two-structure selection inside the water branch and writes the winner into water's own
//! field, so a probe that read whichever component came first read methane's stale zero - and
//! every occupancy printed from it was the other structure's. `cavprwat` is the tell: `2/17`
//! and `1/17`, not `1/23` and `3/23`.

// The oracle values are NeqSim's ground truth, so their full digits are the point.
#![allow(clippy::excessive_precision)]

use azoth_core::AzothError;
use azoth_core::units::{kelvins, pascals};
use azoth_eos::mixture::RootSide;
use azoth_eos::{Cubic, databank, hydrate};

/// `(T in K, P in bara, [f_methane, f_ethane, f_propane, f_water] in bara, the stable
/// structure, and six occupancies as `[small: methane, ethane, propane; large: ...]`)`.
type State = (f64, f64, [f64; 4], usize, [f64; 6]);

const STATES: [State; 3] = [
    (
        293.227_691_103_840,
        100.0,
        [
            73.913_006_100_490_8,
            5.623_707_032_765_60,
            0.757_593_084_435_071,
            0.018_831_198_125_747_6,
        ],
        1,
        [
            0.883_388_125_313_154,
            0.0,
            0.0,
            0.098_687_644_680_376_3,
            0.473_812_737_652_016,
            0.423_505_120_488_215,
        ],
    ),
    (
        288.530_460_404_555,
        50.0,
        [
            39.572_555_564_715_4,
            3.877_475_364_990_52,
            0.635_251_541_845_998,
            0.013_077_400_912_688_7,
        ],
        1,
        [
            0.827_853_003_809_279,
            0.0,
            0.0,
            0.061_464_211_812_117_9,
            0.440_926_974_790_581,
            0.493_360_317_509_043,
        ],
    ),
    (
        297.061_002_920_514,
        200.0,
        [
            134.872_480_485_926,
            7.373_163_724_510_20,
            0.788_096_602_127_644,
            0.026_587_509_996_694_3,
        ],
        1,
        [
            0.923_519_664_857_266,
            0.0,
            0.0,
            0.160_891_597_042_569,
            0.493_336_214_283_545,
            0.341_939_918_179_603,
        ],
    ),
];

/// The reference water phase's fugacity at the capture's first state, in bara.
///
/// NeqSim builds that phase as a pure-water phase of the *host's* class, so it is an SRK water
/// state here - and `reference_water_fugacity` below computes it from this crate's own SRK
/// water rather than taking the number, with this constant asserting the two agree.
const REFERENCE_WATER_FUGACITY_BARA: f64 = 0.018_831_203_388_759_6;

/// The fugacity of pure water at a state, on the cubic the hydrate's fluid runs, in pascals:
/// what NeqSim's `refPhase` contributes to the coefficient.
fn reference_water_fugacity(t: f64, p: f64) -> f64 {
    let (mixture, _) = databank::mixture_of(&["water"], Cubic::Srk, None).expect("water resolves");
    let reduced = mixture
        .reduced_parameters(kelvins(t), pascals(p))
        .expect("the reduction");
    let state = mixture
        .phase_state(&reduced, &[1.0], RootSide::Liquid)
        .expect("a liquid water root");
    state.ln_phi[0].exp() * p
}

/// The four substances' hydrate records, as the resolver builds them.
fn guests() -> Vec<hydrate::HydrateGuest> {
    ["methane", "ethane", "propane", "water"]
        .iter()
        .map(|name| {
            let entry = databank::entry(name, None).expect("the substance is in the databank");
            hydrate::HydrateGuest {
                name: entry.name.clone(),
                langmuir_a: entry.hydrate_langmuir_a,
                langmuir_b: entry.hydrate_langmuir_b,
                former: entry.hydrate_former,
            }
        })
        .collect()
}

#[test]
fn the_occupancies_reproduce_the_capture() {
    for (t, _p_bara, fugacities, structure, expected) in STATES {
        let records = guests();
        // The capture's fugacities are in bara, as NeqSim's internal ones are; the kernel
        // states its own in pascals and converts where the Langmuir product needs it.
        let refs: Vec<f64> = fugacities.iter().map(|value| value * 1.0e5).collect();

        for cavity in 0..2 {
            let occupied = hydrate::occupancy(&records, &refs, structure, cavity, t);
            for (index, guest) in ["methane", "ethane", "propane"].iter().enumerate() {
                let wanted = expected[cavity * 3 + index];
                assert!(
                    (occupied[index] - wanted).abs() < 1.0e-12,
                    "T = {t}: cavity {cavity} {guest} is {}, the capture says {wanted}",
                    occupied[index]
                );
            }
            assert_eq!(occupied[3], 0.0, "water occupies no cage");
        }
    }
}

#[test]
fn the_water_fugacity_coefficient_reproduces_the_capture() {
    // At the answer the coefficient is `alpha_water / P`, because the hydrate's water fugacity
    // *is* the fluid's there - which is the definition of the temperature being solved for.
    for (t, p_bara, fugacities, structure, _) in STATES {
        let records = guests();
        let refs: Vec<f64> = fugacities.iter().map(|value| value * 1.0e5).collect();

        let (found, coefficient) = hydrate::stable_structure(
            &records,
            &refs,
            t,
            p_bara * 1.0e5,
            reference_water_fugacity(t, p_bara * 1.0e5),
        )
        .expect("the sum");
        assert_eq!(found, structure, "T = {t}: the stable structure");
        let expected = fugacities[3] / p_bara;
        assert!(
            (coefficient / expected - 1.0).abs() < 1.0e-6,
            "T = {t}: the coefficient is {coefficient}, and the capture's water fugacity over \
             the pressure is {expected}"
        );
    }
}

/// The cavity sum, asserted on its own rather than only through the coefficient.
///
/// It is `-0.577685053424543` at the first state, against a chemical-potential change of
/// `+0.577684723785035` - they cancel to a part in `1e7`, which is *why* the coefficient comes
/// out as `alpha_w/P` at an answer. A port that dropped a sign in either term would still land
/// near an answer if it dropped the other one the same way, so the term is checked directly.
#[test]
fn the_cavity_sum_reproduces_the_probe() {
    let (t, _p_bara, fugacities, structure, _) = STATES[0];
    let records = guests();
    let refs: Vec<f64> = fugacities.iter().map(|value| value * 1.0e5).collect();

    let mut cavity_sum = 0.0;
    for cavity in 0..2 {
        let occupied: f64 = hydrate::occupancy(&records, &refs, structure, cavity, t)
            .iter()
            .sum();
        cavity_sum += hydrate::CAVITIES_PER_WATER[structure][cavity] * (1.0 - occupied).ln();
    }
    assert!(
        (cavity_sum + 0.577_685_053_424_543).abs() < 1.0e-12,
        "the cavity sum is {cavity_sum}, and the probe prints -0.577685053424543"
    );
}

#[test]
fn a_full_cavity_is_refused_rather_than_clamped() {
    let records = guests();
    // Fugacities large enough that the cavity sum passes one: at 100 bar the guests' are of
    // order `1e7` Pa and the Langmuir constants of order one, so `1e30` saturates.
    let saturated = vec![1.0e30, 1.0e30, 1.0e30, 1.0e30];
    let error = hydrate::stable_structure(
        &records,
        &saturated,
        293.0,
        1.0e7,
        reference_water_fugacity(293.0, 1.0e7),
    )
    .expect_err("a fully occupied cavity has no logarithm");
    assert!(matches!(error, AzothError::OutOfRange { .. }), "{error:?}");
}

/// This crate's pure-water SRK fugacity reproduces NeqSim's reference phase.
///
/// The coefficient's third term is `ln(f_w^ref/f_w^fluid)`, and `f_w^ref` is a *computed*
/// quantity rather than a constant: NeqSim builds it from the host's own equation. So the two
/// implementations have to agree on it, and this is where the capture's printed
/// `ref_water_fugacity` is held to this crate's water.
#[test]
fn the_reference_water_fugacity_reproduces_the_probe() {
    let (t, p_bara, _, _, _) = STATES[0];
    let p = p_bara * 1.0e5;
    let computed = reference_water_fugacity(t, p) / 1.0e5;
    assert!(
        (computed / REFERENCE_WATER_FUGACITY_BARA - 1.0).abs() < 1.0e-6,
        "the reference water fugacity is {computed} bar, and the probe prints \
         {REFERENCE_WATER_FUGACITY_BARA}"
    );
}
